### ISSUE TITLE
Base.unzip()

### ISSUE THREAD

**Issue by ZacCranko** [JuliaLang/julia#13942]
Hi there,

apologies if this has already been addressed somewhere, but is there a reason that there is no `unzip()` function in Base?

Ideally this would be a function that would take a `Vector{Tuple{ ... }}` and return a `Tuple{Vector, ..., Vector}` for output. E.g.
```
julia> v = [(1,"a",:meow), (2,"b",:woof), (3,"c",:doh!)]; unzip(v)
([1,2,3],ASCIIString["a","b","c"],[:meow,:woof,:doh!])
```

A naive implementation might be something like
```julia
function unzip(input::Vector)
    n = length(input)
    types  = map(typeof, first(input))
    output = map(T->Vector{T}(n), types)

    for i = 1:n
       @inbounds for (j, x) in enumerate(input[i])
           (output[j])[i] = x
       end
    end
    return (output...)
end
```

**Comment by StefanKarpinski** [comment]
We definitely need an `unzip` function in Base.

**Comment by ivarne** [comment]
Or even simpler

```julia
unzip(a) = zip(a...) 
```

**Comment by yuyichao** [comment]
[No](https://github.com/JuliaLang/julia/issues/13930#issuecomment-155142306)

**Comment by StefanKarpinski** [comment]
Trying the zip(...) thing on any non-toy problem shows that it's not so good pretty fast, unfortunately.

**Comment by tlnagy** [comment]
I just ran into this issue when I absentmindedly used the Python-esque `zip(arr...)` where `length(arr)=6000`. It was not pretty. A dedicated `unzip` would be nice. Maybe even throw a warning that you shouldn't splat anything nontrivial.

**Comment by yuyichao** [comment]
> Maybe even throw a warning that you shouldn't splat anything nontrivial.

Like almost all other performance tips in julia, doing so is perfectly fine and sometimes wanted for various reason. (OK, splatting 6000 elements is a bit crazy, but splatting unknown size can be useful....). The only thing that matters is to not put it in performance critical path.

It looks like this is not covered in the performance tip?

**Comment by tlnagy** [comment]
> Like almost all other performance tips in julia, doing so is perfectly fine and sometimes wanted for various reasons. (OK, splatting 6000 elements is a bit crazy, but splatting unknown size can be useful....). The only thing that matters is to not put it in performance critical path.

Yeah, a warning is a bad idea.

> It looks like this is not covered in the performance tip?

I think this should also be marked under differences from other languages. People coming from Python are probably used to the `zip(*arr)` syntax and `zip(arr...)` might be a tempting target.

**Comment by hayd** [comment]
You can dispatch only on arrays of tuples:

    function unzip{T<:Tuple}(A::Array{T})
        res = map(x -> x[], T.parameters)
        res_len = length(res)
        for t in A
            for i in 1:res_len
                push!(res[i], t[i])
            end
        end
        res
    end

**Comment by StefanKarpinski** [comment]
Bump, we still need and unzip function.

**Comment by StefanKarpinski** [comment]
BUMP if anyone is bored...

**Comment by JeffreySarnoff** [comment]
```julia

function unzip(zipped::Base.Iterators.Zip2{Array{T1,1}, Array{T2,1}}) where {T1,T2}
    n = length(zipped)
    vec1 = Vector{T1}(n)
    vec2 = Vector{T2}(n)
    i = 1
    for pair in enumerate(zipped)
        vec1[i], vec2[i] = pair[2]
        i = i+1
    end
    return vec1, vec2
end
```

**Comment by JeffreySarnoff** [comment]
maybe <= 1.5x is
```julia
function unzip(zipped::Base.Iterators.Zip2{Array{T1,1}, Array{T2,1}}) where {T1,T2}
    return map(first,zipped), map(last,zipped)
 end
```

**Comment by martinholters** [comment]
```julia
unzip(zipped::Base.Iterators.Zip2) = (zipped.a, zipped.b)
```
But all those differ in how they treat `length(a) != length(b)`. I think I'd favor the behavior of the first.

This trivial case apart, the desired semantics are not absolutely clear to me: Given an iterator `itr`, should `unzip(itr)` return an iterator `unzipped`, that generates iterators `unzipped[1]` to `unzipped[N]`, where `N` is the length of the iterators generated by `itr`? What if the first 10000 values generated by `itr` are `Tuple`s of length `N==3`, but then comes a `Tuple` of length 2? Throw when iterating at least one `unzipped[n]` up to that point? Or eagerly collect `itr` the moment `unzip` is called?

Think about
```julia
unzip(f(n) for n in Base.Iterators.countfrom(0))
```
for different `f`s to get an idea of my open questions.

**Comment by spalato** [comment]
There seems to be multiple use cases for an unzip, with different requirements.

A common case is involves Arrays of Tuples, in which the size is known and array shape must be preserved. Some suggestions here do not obey this constraint.

I have come up with a solution for this specific case using `Base.Cartesian`. It's up to twice as fast as the alternative commonly suggested alternative for NTuples, and more than 10x for heterogeneous tuples.
See: https://github.com/spalato/Destruct.jl
Feel free to use and steal.

**Comment by bramtayl** [comment]
Can we get this for generalized iterators and not just arrays? Basically just a collect_many modeled after collect?

**Comment by bramtayl** [comment]
I could give it a stab? The trickiest bit (I think) will be getting an equivalent of default_eltype for zero length eltype unknown iterators

**Comment by StefanKarpinski** [comment]
Go for it, @bramtayl!

**Comment by bramtayl** [comment]
Tada, this seems to work. Great excuse not to finish my grading.

```julia
struct Unzip{Iterator}
    iterator::Iterator
end

iterate(iterator::Unzip) = iterate(iterator.iterator)
iterate(iterator::Unzip, state) = iterate(iterator.iterator, state)

IteratorEltype(iterator::Unzip) = IteratorEltype(iterator.iterator)
eltype(iterator::Unzip) = eltype(iterator.iterator)

IteratorSize(iterator::Unzip) = IteratorSize(iterator.iterator)
axes(iterator::Unzip) = axes(iterator.iterator)
size(iterator::Unzip) = size(iterator.iterator)
length(iterator::Unzip) = length(iterator.iterator)

struct ModelArray{ElementType, NumberOfDimensions, Model, Rest <: Tuple} <:
    AbstractArray{ElementType, NumberOfDimensions}
    model::Model
    rest::Rest
end

model_array(model, rest...) =
    ModelArray{
        Tuple{eltype(model), eltype.(rest)...},
        ndims(model),
        typeof(model),
        typeof(rest)
    }(model, rest)

axes(array::ModelArray) = axes(array.model)
size(array::ModelArray) = size(array.model)

IndexStyle(array::ModelArray) = IndexLinear()

@propagate_inbounds getindex(array::ModelArray, index::Int) = (
    getindex(array.model, index),
    getindex.(array.rest, index)...
)

@propagate_inbounds setindex!(array::ModelArray, value::Tuple, index::Int) = (
    setindex!(array.model, value[1], index),
    setindex!.(array.rest, tail(value), index)...
)

push!(array::ModelArray, value::Tuple) = (
    push!(array.model, value[1]),
    push!.(array.rest, tail(value))
)

# TODO: use fieldtypes instead of value_field_types in a future with more constant propagation
value_field_types(a_type) =
    ntuple(index -> Val(fieldtype(a_type, index)), fieldcount(a_type))

function similar(
    array::Unzip,
    ::Type{ElementType},
    dims::Dims
) where {ElementType, NumberOfDimensions}

    inner_array(::Val{InnerElementType}) where InnerElementType =
        Array{InnerElementType}(undef, dims)
    model_array(inner_array.(value_field_types(ElementType))...)
end

# Unzip needs to replicate AbstractArray similar machinery
similar(a::Unzip, ::Type{T}) where {T}                     = similar(a, T, to_shape(axes(a)))
similar(a::Unzip{T}, dims::Tuple) where {T}                = similar(a, T, to_shape(dims))
similar(a::Unzip{T}, dims::DimOrInd...) where {T}          = similar(a, T, to_shape(dims))
similar(a::Unzip, ::Type{T}, dims::DimOrInd...) where {T}  = similar(a, T, to_shape(dims))
similar(a::Unzip, ::Type{T}, dims::Tuple{Union{Integer, OneTo}, Vararg{Union{Integer, OneTo}}}) where {T} = similar(a, T, to_shape(dims))

Tuple{Int64} <: Tuple{Vararg{Int64}}
similar(array::ModelArray, ::Type{ElementType}, dims::Dims) where ElementType =
    similar(Unzip(nothing), ElementType, dims)

_similar_for(c::Unzip{Nothing}, T, itr, ::SizeUnknown) =
    similar(c, T, 0)
_similar_for(c::Unzip{Nothing}, T, itr, ::HasLength) =
    similar(c, T, Int(length(itr)::Integer))
_similar_for(c::Unzip{Nothing}, T, itr, ::HasShape) =
    similar(c, T, axes(itr))

collect(iterator::Unzip) = _collect(
    Unzip(nothing),
    iterator,
    IteratorEltype(iterator),
    IteratorSize(iterator)
)

# Examples

Generator(x -> (x, x + 1.0), [1]) |> Unzip |> collect
Generator(x -> (x, x + 1.0), [1, missing]) |> Unzip |> collect
zip([1], [1.0]) |> Unzip |> collect
[(1, 1.0)] |> Unzip |> collect
Filter(x -> x[1] == x[2], [(1, 1.0)]) |> Unzip |> collect
``

**Comment by bramtayl** [comment]
The above strategy turns out to be extremely sub-optimal for EltypeUnknown iterators, but it still works. I'm not sure I have the programming wherewithall to design a better system.

**Comment by bramtayl** [comment]
I'm thinking it would probably just need specialized ::ModelArray methods for grow_to! and collect_to! to be efficient, I'll see what I can cook up

**Comment by bramtayl** [comment]
I'm leaving the code here for now https://github.com/bramtayl/Unzip.jl

**Comment by bramtayl** [comment]
It should only work on master

**Comment by oxinabox** [comment]
Can this become a PR, and then get merged?

**Comment by bramtayl** [comment]
I'm not sure I have the programming chops to make a PR. Or rather, I have some questions: what file should the code go into? What modifications do I need to make to take into account bootstrap order? Do I need boundschecking in the ModelArray constructor to make sure all the inputs are the same size? How can I adjust the code to take into account offset axes? What would be a good way to effectively benchmark this? How can I tell if the widening optimizations are working as intended?

**Comment by bramtayl** [comment]
Also, I've run into a bit of a problem. Not sure what do do in the case of EltypeUnknown, SizeUnknown iterators where `@default_eltype` yields `Union{}`. We need to know at least the length of the tuple to start unzipping...

**Comment by mschauer** [comment]
I think it is okay to make some assumption on the iterator which is to be unzipped (e.g. not returning tuples of different length) and  it would okay to give an error message in the case this contract is not fulfilled.

**Comment by o314** [comment]
Well, well. take [that in python](https://stackoverflow.com/questions/19339/transpose-unzip-function-inverse-of-zip), write this in julia:

```julia
s = [(1, 'a'), (2, 'b'), (3, 'c')]

@test collect(zip(zip(s)...)...) == s
Test Passed
```
so

```julia
unzip(z) = collect(zip(z...)...)

@test (unzip ∘ zip)(s) == s
Test Passed
```

One more impl, quickly usable.
Let's call it the haiku version of julia unzip

**Comment by o314** [comment]
Well, well. take [that in python](https://stackoverflow.com/questions/19339/transpose-unzip-function-inverse-of-zip),
write this in julia:


```julia
s1,s2 = (1,2,3),('a','b','c')
s = tuple(zip(s1,s2)...)

@test tuple(zip(s...)...) == (s1,s2)
Test Passed
```
so

```julia
unzip(z) = tuple(zip(z...)...)

@test (unzip ∘ zip)(s1,s2) == (s1,s2)
Test Passed
```

One more impl, quickly usable.
Let's call it the haiku version of julia unzip


*PS: edited v1 w array removed, was not type stable. v2 w tuple seems ok*

**Comment by StefanKarpinski** [comment]
We have an implementation; all that's required is for someone to make a PR to add it to Base. The issue is already tagged as "good first issue" and "help wanted" so hopefully someone feels like trying it.

**Comment by o314** [comment]
@KristofferC

Sorry, code was so short it pass under my radar

```julia
using Test
unzip(z) = tuple(zip(z...)...)

s,t = [1:1_000_000...],[1:1_000_000...];
z = zip(s,t) |> collect;
@test_throws StackOverflowError unzip(z)

s,t = [1:100...],[1:100...];
z = zip(s,t) |> collect;
@time unzip(z);
# 4.105195 seconds (26.41 M allocations: 1.021 GiB, 17.28% gc time)
```
it's an useless toy

**Comment by StefanKarpinski** [comment]
Yes, that's [been noted](https://github.com/JuliaLang/julia/issues/13942#issuecomment-155761636):

> Trying the `zip(...)` thing on any non-toy problem shows that it's not so good pretty fast, unfortunately.

I was referring to @bramtayl's Unzip package, which has a real implementation... but also appears to have been deleted from GitHub. So I guess we do not actually have an implementation.

**Comment by o314** [comment]
Of course, that's was clearly understandable.
I was just answering to Kristoffer about the splat option.
Didn't figure numbers were so bad.

Thanks for pointing out the correct track.
Still a bit hard for me at this point to contribute to Base.
Good luck to other people...

**Comment by StefanKarpinski** [comment]
No worries. I didn't bother measuring the splatting option because I knew it was going to be terrible 😬

**Comment by o314** [comment]
OK. That is surprisingly pretty good however in python
```python
def unzip(z): return zip(*z)

def bench_zip_unzip():
    z = [*zip(range(1_000_000), range(1_000_000))]
    u = [*unzip(z)]

import timeit
timeit.timeit(bench_zip_unzip, number=1)
# 0.4816088059997128 second
```

**Comment by simonbyrne** [comment]
one other thing that would be useful would be to allow it to take a function first argument, so that you could do e.g.
```
s,c = unzip(sincos, x)
```
for a vector (or array) `x`.

**Comment by StefanKarpinski** [comment]
Copying some thoughts from a thread on Slack, the utility of `unzip` is that it works like `collect` does on a single iterable, materializing it into a concrete vector but producing as many vectors as there are elements in each tuple that the iterable yields. The iterable _must_ yield the same sized tuple each time; it need not, however, yield the same element types each time; `unzip` should handle this the same way that collect or comprehensions do (they're slightly different, so I guess that's a design choice), for each vector that it outputs. So the result of `unzip(itr)` should be the same as
```jl
ntuple(length(first(itr))) do i
    [t[i] for t in itr]
end
```
except that it only consumes `itr` a single time and it should error if any of the tuples has a different length than the first one. That will be a bit tricky to write 🤔. Mocked up example:
```jl
julia> itr = [(1,2,3),(3,1.5,'x')]
2-element Array{Tuple{Int64,Real,Any},1}:
 (1, 2, 3)
 (3, 1.5, 'x')

julia> unzip(itr)
([1, 3], Real[2, 1.5], Any[3, 'x'])
```

**Comment by bramtayl** [comment]
See #30987

**Comment by StefanKarpinski** [comment]
Thanks, @bramtayl!

**Comment by o314** [comment]
It will be good if we can unzip `Array{Pair}` too

eg. `unzip([k=>v for (k,v) in zip('a':'c',1:3)]) == unzip(zip('a':'c',1:3))`

**Comment by JeffreySarnoff** [comment]
bump

**Comment by bramtayl** [comment]
https://gist.github.com/bramtayl/2d6d323582f7c3158f9a1b5299997d68

Not skilled enough to make a PR, but I think this should be a pretty robust implementation

**Comment by zzyxyzz** [comment]
Would this be considered an acceptable solution?

```julia
unzip(z::Iterators.Zip) = z.is
unzip(itr) = unzip(collect(itr))
unzip(::AbstractArray) = error("Can only unzip array of tuples.")

@generated function unzip(a::AbstractArray{T}) where {T<:Tuple}
    types = T.types
    types[end] <: Vararg && error("Tuples in the array should have the same length.")
    isempty(types) && return :(())
    n = length(types)
    quote
        @nexprs $n i -> out_i = similar(a,$types[i])
        @inbounds for j in eachindex(a)
            @nexprs $n i -> out_i[j] = a[j][i]
        end
        return @ntuple $n out
    end
end
```

90% of this implementation is taken from the package [Destruct.jl](https://github.com/spalato/Destruct.jl) by @spalato, who [commented](https://github.com/JuliaLang/julia/issues/13942#issuecomment-374378742) earlier in this thread.

This implementation is maximally efficient for dense arrays of tuples (which was the motivating example for this issue). It is slightly less efficient for arrays with non-standard indexing, due to reliance on `eachindex`.

For the `Zip` iterator, it just returns the underlying iterators *without* truncating them to a common size. I.e. `a,b,c = unzip(zip(a,b,c))` is a no-op.

For other iterators and generators, it first collects the output into an array and unzips that. This is not ideal, since it requires twice as much memory, but I simply don't have enough experience with iterator internals to implement it better. (This is also why I haven't tried to ressurect @bramtayl's implementation: I simply couldn't wrap my head around it yet :smile:).

AFAICT, the above code fulfils all the functional requirements set out by @StefanKarpinski [above](https://github.com/JuliaLang/julia/issues/13942#issuecomment-461486754). If this implementation is deemed acceptable, despite using generated functions and sacrificing efficiency in some cases, I can add tests and documentation and open a PR.

**Comment by bramtayl** [comment]
The code in the gist was written relatively recently, so I think it should be copy-pastable? It sounds like its a lot more fully featured than what your proposing (works with arbitrary length iterators, returning a variable number of items). Just need someone to figure out the PR details (something I never seem to be able to do successfully)

**Comment by bramtayl** [comment]
I'm also happy to explain whatever is helpful. Lispy-tuple-recursion and iterative widening are both very useful to know how to do regardless.

**Comment by zzyxyzz** [comment]
> It sounds like its a lot more fully featured than what your proposing (works with arbitrary iterators, returning a variable number of items).

Iterators are supported, but with some overhead. I think the main difference between the two variants is that you produce arrays with `missing` values if the tuples have different length. Stefan's [comment](https://github.com/JuliaLang/julia/issues/13942#issuecomment-461486754) implies that an error should be thrown in this case, but I can see how using `missing` values could be sometimes more useful. If this behavior is preferred, I think it can be implemented in the generating-function-based approach as well.

> I'm also happy to explain whatever is helpful. Lispy-tuple-recursion and iterative widening are both very useful to know how to do regardless.

Thanks, I might have to take you up on that offer yet :smile:. But I really don't feel qualified to finish what you started there. I just spent several hours trying to understand your implementation and still only have a general idea of how it works. And that's without even touching on the `missing` and widening stuff.

I also have to ask, why is this style preferred in Julia (in array library code, at least)? As far as I can tell, the end result is that a specialized method is produced for every combination of tuple and container type that is invoked in user code. Which is more or less the same as what a generated function would do. On the other hand, a generated function does exactly what it advertises, while the other solution goes back and forth through twenty layers of abstraction (this is not a metaphor; I think the call chain really is this deep, even though it all gets inlined in the end). I'm sorry if this sounds argumentative, and I'm also starting to think this may be the wrong place to ask such questions, but I'm genuinely curious why layered abstractions relying on vararg splatting on every other step are preferred to gen-funcs in general.

**Comment by bramtayl** [comment]
Two things:

1) Collecting into a array first is unworkable when some of the columns are small unions, because the eltype will be Any
2) Because Julia is a dynamic language, there is no return type. So you can't look at an iterator and guess what types each of the columns will end up being, or for that matter, how many columns there are going to be in the first place.

**Comment by zzyxyzz** [comment]
> 1. Collecting into a array first is unworkable when some of the columns are small unions, because the eltype will be Any

You're right about that, of course. Though I think you are holding yourself to a higher standard than the standard library (no pun intended). Maybe it would be preferable to use your proposed changes to improve `collect`, so that it doesn't widen types more than necessary? Then less skilled people, such as myself, could get away with falling back on it without too high a cost :smile:.

Also, if the "iterator" passed to `unzip` is an actual array, there is the same problem: `[(1,"a"), (2,3)]` infers as `Array{Tuple{Int64,Any},1}`. Does your implementation tighten the output type to `Tuple{Array{Int64,1}, Array{Union{Int64,String},1}}` in this case as well?

**Comment by bramtayl** [comment]
Yes

**Comment by bramtayl** [comment]
Sorry, I didn't mean for that to sound short. I really appreciate all the work you're doing, and happy to collaborate on any solutions.

**Comment by zzyxyzz** [comment]
That's impressive. Can you explain, in general terms, how you do it? In my understanding that would require backtracking and converting some of the columns when you encounter a tuple with an unexpected length or type. And the result type could not be determined until the operation is complete, no?

**Comment by bramtayl** [comment]
Yes and yes. `push_widen` and `setindex_widen_up_to` already do this in collect for Base (and I outsource to them). Whether or not the result type can be determined depends on how smart inference is being that day, but the short answer is that because Julia is a dynamic language, it doesn't matter anyway.

**Comment by bramtayl** [comment]
But I do have a doctest for inference:

```
julia> stable(x) = (x, x + 0.0, x, x + 0.0, x, x + 0.0);

julia> @inferred unzip(Generator(stable, 1:4))
```

So, in some cases, inference can figure out what's going on just fine.

**Comment by zzyxyzz** [comment]
But in general, this is not type-stable, I guess, especially with arrays rather than generators?

**Comment by bramtayl** [comment]
Yes. Which is ok, that's par for the course in Julia.

**Comment by zzyxyzz** [comment]
Thanks for the explanations! Gives me a lot to think about.

Last question: Do you know why this behavior is not implemented for `collect`? Is it just that nobody got around to it yet, or is there some other hidden cost to consider?

**Comment by bramtayl** [comment]
I'm not sure I'm qualified to answer that question, but I think #25553 is the place to start looking

**Comment by bramtayl** [comment]
The answer is here https://github.com/JuliaLang/julia/pull/26501#issuecomment-373957902 I think

**Comment by zzyxyzz** [comment]
Hmm... correct me if I'm wrong, but [#26501 (comment)](https://github.com/JuliaLang/julia/pull/26501#issuecomment-373957902) and [#26501 (comment)](https://github.com/JuliaLang/julia/pull/26501#issuecomment-373961020) seems to suggest that diverging tuple fields widening to `Any` rather than `Union` is by design (because compile times). So the tradeoff here is partly in how much we want to tax the compiler.

@StefanKarpinski, could you chime in with an opinion?

**Comment by o314** [comment]
Notice that there is some mismatch between what we expect from iterator protocol v1

![annex - unzip as iterators dot](https://user-images.githubusercontent.com/4026002/66010368-ed707480-e4be-11e9-8d22-65239b37796c.png)

And how unzip works internally

![annex - unzip as parallelized-iterator dot](https://user-images.githubusercontent.com/4026002/66010468-548e2900-e4bf-11e9-845d-bc6cc8c19cb3.png)


Without seekable io / indexable iterator and/or buffering, that could never fit.

**Comment by bramtayl** [comment]
I took at stab at simplifying the code in my gist. I made a little headway but not much. If it's any comfort, many of the functions (map_unrolled, partial_map, zip_missing) would fit nicely into a unexported mini-module for unrolling tuple operations. I've been kicking around a plan for that for a long time.

**Comment by oxinabox** [comment]
It would be really nice to get this solved for 1.6.

**Comment by JeffreySarnoff** [comment]
@StefanKarpinski I remember that you had a full grasp of this -- is anything of deep import holding it back?

**Comment by StefanKarpinski** [comment]
Only performance, Iirc.

**Comment by zzyxyzz** [comment]
@JeffreySarnoff, I may be misremembering some things, but I think there were some unresolved questions aside from performance:

1. What to do if tuples have different length?
   * Fill in `missing`s?
   * Throw error?
2. What to do if the same field has different types in different tuples?
   * Error?
   * Promote to `Any`?
   * Promote to smallest possible `Union`?
   * Promote to parent type?
   * Widen numeric types?
3. Lazy or strict semantics?
   * In particular, do we want `unzip` of `zip` be the identity? (Rather than throwing an error if the input arrays are of different size, as `zip` currently does). 

Maybe some other things too, but it's been a while since I read through the entire thread.

Note that these issues only really arise in `TypeUnknown` iterators. For homogeneous tuple arrays there is the [Destruct](https://github.com/spalato/Destruct.jl) package, which implements a simple and efficient `unzip` operation.

**Comment by bramtayl** [comment]
#33324 seems pretty good to me still. If we want to simplify the code I think the first step would be to solve #31909

**Comment by bramtayl** [comment]
I wrote up https://github.com/bramtayl/Unzip.jl and it should be registered 3 days from now.

**Comment by JeffreySarnoff** [comment]
_iknow uknow this is furiously obvious_
- `unzip` that which is `zip`ped  
```julia
	unzip(x::Base.Iterators.Zip) = z.is
```

**Comment by bramtayl** [comment]
Feel free to open up a PR over there, but I'm so sure about that method. The behavior in my version `unzip` is to collect new vectors, not pass along old ones. `z.is` also pretty easy to type if you already have a zip.

**Comment by zzyxyzz** [comment]
> _iknow uknow this is furiously obvious_
> 
>     * `unzip` that which is `zip`ped
> 
> 
> ```julia
> 	unzip(x::Base.Iterators.Zip) = z.is
> ```

How is this obvious? It's *one possible way* to define `unzip` of `zip`, and has the (IMO major) disadvantage that `unzip` is not guaranteed to return arrays of equal length.

**Comment by JeffreySarnoff** [comment]
(it works better with `z.is` replaced by `x.is`)
It will return the arrays that were given to `zip`.
How is that a major disadvantage for `unzip`? 
I conceived that its essential purpose is to "undo" what `zip` did.

**Comment by JeffreySarnoff** [comment]
@zzyxyzz I meant "obvious" in the sense that this is what `dump(zip(a,b))` shows. I had not known about the inner structure of our zips until today .. so, "afaic a clear candidate once the internal representation is reviewed" is how I have revised that phrase (above).

**Comment by mschauer** [comment]
> return arrays of equal length

good point

**Comment by bramtayl** [comment]
Maybe `parents` would be a good name for what `function (x::Base.Iterators.Zip) = x.is end does

**Comment by zzyxyzz** [comment]
@JeffreySarnoff, I'm not saying that making `unzip(zip(...))` a no-op is necessarily a bad idea. It's just that `unzip` has a surprising number of loose ends and subtle trade-offs for something that is so simple in principle. This is marked as "good first issue", but it really isn't. Over the years several people have taken a stab at it (@bramtayl has come furthest), but it's no accident that there is still no generally accepted solution.

**Comment by zzyxyzz** [comment]
Maybe this issue should be renamed, following Julia tradition, "Taking unzip seriously". :stuck_out_tongue:

**Comment by bramtayl** [comment]
Registered!

**Comment by JeffreySarnoff** [comment]
For `unzip(x::Base.Iterators.Zip)` where the lengths of `x.is` are equal, would you rather return `x.is` or a copies of the constituents?

**Comment by zzyxyzz** [comment]
@JeffreySarnoff

> For unzip(x::Base.Iterators.Zip) where the lengths of x.is are equal, would you rather return x.is or a copies of the constituents?

My personal take on this:

In any decent implementation, `unzip` should call its source iterator exactly once for each input tuple. This forces `unzip` to construct the output arrays in lock-step; it cannot finish the first array before starting on the second. Thus, `unzip` cannot act as a lazy iterator in any reasonable sense. It is fundamentally a `collect`-like operation. So the safe, consistent thing to do would be to emulate the behavior of `collect` and **always return freshly allocated arrays**.

But I do admit that the temptation to special-case `unzip(::Base.Iterators.Zip)` is strong. And it wouldn't be the first time that safe, predictable behavior is sacrificed for performance in Julia. :smile:

**Comment by o314** [comment]
No post in 2021 ? So let's give it a try now! HNY 2022 Unzip !

# Multidispatch-friendly zen of unzip in julia

> @ https://gist.github.com/o314/214e26c6fb70512b56597d633dd87e6f  
> see https://github.com/JuliaLang/julia/issues/13942

OK `unzip(a) = zip(a...)` fails in Julia
But Zen of python is great - ok, it's somewhat a lie, it's also great for julia !
So let's try to bring correct simple things first, and (maybe) complicated ones later or away.

FWIW, my good-enough-poor-man, but usable in prod, isn't it, unzip considering than

- writing 5 lines > waiting 5 years
- not optimized for nonexistant use case of unzip (for streaming or tensorflow or whatever)
- third party pkg for 5 lines is a joke or an industrial hazard (software bom somebody?)
- con : 5 sloc of src; 50 lines of test. so what?



```julia
using Test
import Base.Iterators as _I

unzip(s...) = unzip(collect(s))
unzip(vs::Vector{<:Vector}) =
    let M=length(vs), N=mapfoldl(length, min, vs); # todo remove me when SVector is in Base
        ([vs[i][j] for i in 1:M] for j in 1:N)
    end
unzip(a::Vector{<:Pair}) = [k for (k,_) in a], [v for (_,v) in a]
```

## TEST

```julia
import Base.Iterators as _I
using Test
# zipdata(M,N) = let v=collect(1:M), vt=ntuple(N) do _; copy(v) end; vt end
data(M,N) = ntuple(M) do i; fill(i,N) end
data(N) = let ks=_I.take(_I.cycle('a':'z'), N), vs=(1:N...,); (k=>v for (k,v) in zip(ks,vs)) end
```

### VALIDITY TEST

```julia
# unzip of vector
@test data(5,3) == ([1,1,1],[2,2,2],[3,3,3],[4,4,4],[5,5,5])
@test unzip(data(5,3)...) |> collect == ([1,2,3,4,5],[1,2,3,4,5],[1,2,3,4,5]) |> collect

# unzip of pair vector
@test data(5) |> collect == ('a'=>1, 'b'=>2, 'c'=>3, 'd'=>4, 'e'=>5) |> collect
@test unzip(data(5) |> collect) |> collect == (['a','b','c','d','e'], [1,2,3,4,5]) |> collect
```

### SPEED TEST

```
# unzip of vector
julia> @time unzip(data(1000,3)...);
  0.029086 seconds (42.07 k allocations: 2.766 MiB, 99.28% compilation time)

julia> @time unzip(data(1_000_000,3)...);
  1.507531 seconds (4.04 M allocations: 223.797 MiB, 18.21% gc time, 72.32% compilation time)

julia> @time unzip(data(1_000_000,3)...);
  0.294386 seconds (1.00 M allocations: 152.588 MiB)

julia> @time unzip(data(1000,50)...);
  0.000727 seconds (1.01 k allocations: 531.922 KiB)

julia> @time unzip(data(1_000_000,50)...);
  1.082615 seconds (1.00 M allocations: 518.799 MiB, 48.09% gc time)

julia> @time unzip(data(1_000_000,50)...);
  0.527460 seconds (1.00 M allocations: 518.799 MiB)

# unzip of pair vector
julia> @time unzip(data(1000));
  2.728774 seconds (166.12 k allocations: 10.524 MiB, 99.98% compilation time)

julia> @time unzip(data(1000));
  0.000334 seconds (2.00 k allocations: 116.375 KiB)

julia> @time unzip(data(1_000_000) |> collect);              # BUG wo collect
  0.634841 seconds (3.50 M allocations: 178.888 MiB, 18.39% gc time, 57.41% compilation time)
```

**Comment by complyue** [comment]
Per 1.7.2, the `zip(...)` trick would not even scale to 1k!

```
$ julia --version
julia version 1.7.2
$ time julia -e 'zip(collect([3,2,5] for _ in 1:10)...)|>collect' 
julia -e 'zip(collect([3,2,5] for _ in 1:10)...)|>collect'  0.38s user 0.09s system 98% cpu 0.474 total
$ time julia -e 'zip(collect([3,2,5] for _ in 1:1000)...)|>collect'
ERROR: StackOverflowError:
Stacktrace:
 [1] _zip_iterate_interleave(xs1::NTuple{980, Tuple{Int64, Int64}}, xs2::Tuple{}, ds::NTuple{980, Missing})
   @ Base.Iterators ./iterators.jl:368
 [2] _zip_iterate_interleave (repeats 20 times)
   @ ./iterators.jl:369 [inlined]
 [3] _zip_iterate_all(is::NTuple{1000, Vector{Int64}}, ss::NTuple{1000, Tuple{}})
   @ Base.Iterators ./iterators.jl:354
 [4] iterate
   @ ./iterators.jl:340 [inlined]
 [5] copyto!(dest::Vector{NTuple{1000, Int64}}, src::Base.Iterators.Zip{NTuple{1000, Vector{Int64}}})
   @ Base ./abstractarray.jl:890
 [6] _collect(cont::UnitRange{Int64}, itr::Base.Iterators.Zip{NTuple{1000, Vector{Int64}}}, #unused#::Base.HasEltype, isz::Base.HasShape{1})
   @ Base ./array.jl:655
 [7] collect(itr::Base.Iterators.Zip{NTuple{1000, Vector{Int64}}})
   @ Base ./array.jl:649
 [8] |>(x::Base.Iterators.Zip{NTuple{1000, Vector{Int64}}}, f::typeof(collect))
   @ Base ./operators.jl:966
julia -e 'zip(collect([3,2,5] for _ in 1:1000)...)|>collect'  130.60s user 3.81s system 99% cpu 2:14.47 total
```

Hell I miss an `unzip`!

**Comment by ParadaCarleton** [comment]
Someday `unzip` will work. Eventually :sweat_smile:

**Comment by adienes** [comment]
note that since https://github.com/JuliaLang/julia/pull/50435 was merged, there is precedent to throw on follow-up operations when zipped iterators have unequal lengths. so many of the previous concerns upthread about the desire to & controversy of defining `unzip(z::Zip) = zip.is` are addressed & no longer super relevant

**Comment by jakewilliami** [comment]
I've just caught up on the years of work pertaining to the `unzip` function!  Looks like it's almost ready to ship in #33515.

**Comment by StefanKarpinski** [comment]
Yeah, this seems like the perfect being the enemy of the good enough. My PR worked fine and has a sane, legible implementation. It merely isn't the fastest possible—or at least didn't at the time, it may have gotten better since then. I think we should just merge the pretty good implementation.

**Comment by el-oso** [comment]
> Yeah, this seems like the perfect being the enemy of the good enough. My PR worked fine and has a sane, legible implementation. It merely isn't the fastest possible—or at least didn't at the time, it may have gotten better since then. I think we should just merge the pretty good implementation.

Agreed ... when are you merging this (just pick one)? It has been 10 years ... maybe it is time.
